File: src/welch/brainmess/Instruction.java
package welch.brainmess;

/**
 * Represents a Brain Mess Instruction.
 * @author Michael Welch
 *
 */
public enum Instruction
{
	/**
	 * Represents the instruction that moves the tape forward one cell.
	 */
	MoveForward
	{
		public void execute(ExecutionContext context) { context.moveForward(); }
	},
	
	/**
	 * Represents the instruction that moves the tape backward one cell.
	 */
	MoveBackward
	{
		public void execute(ExecutionContext context) { context.moveBackward(); }
	},
	
	/**
	 * Represents the instruction that increments the value of the current cell.
	 */
	Increment
	{
		public void execute(ExecutionContext context) { context.increment(); }
	},
	
	/**
	 * Represents the instruction that decrements the value of the current cell.
	 */
	Decrement
	{
		public void execute(ExecutionContext context) { context.decrement(); }
	},
	
	/**
	 * Represents the instruction that inputs a character and stores its integer
	 * value on the tape at the current location.
	 */
	Input
	{
		public void execute(ExecutionContext context) { context.input(); }
	},
	
	/**
	 * Represents the instruction that reads the value of the current cell and outputs
	 * the character it represents.
	 */
	Output
	{
		public void execute(ExecutionContext context) { context.ouput(); }
	},
	
	/**
	 * Represents the test and jump forward instruction.
	 */
	TestAndJumpForward
	{
		public void execute(ExecutionContext context) { context.testAndJumpForward(); }
	},
	
	/**
	 * Represents the test and jump backward instruction.
	 */
	TestAndJumpBackward
	{
		public void execute(ExecutionContext context) { context.testAndJumpBackward(); }
	},
	
	/**
	 * A no operation instruction. Used for all non-instruction characters in the program.
	 */
	NOP
	{
		public void execute(ExecutionContext context) { }
	};
	
	
	/**
	 * Executes the current instruction.
	 * @param context The execution environment in which to execute the instruction.
	 */
	public abstract void execute(ExecutionContext context);
	
	/**
	 * Parses an Instruction from the specified character value.
	 * @param value
	 * @return
	 */
	public static Instruction parseInstruction(char value)
	{
		switch(value)
		{
		case '>':
			return Instruction.MoveForward;
			
		case '<':
			return Instruction.MoveBackward;
			
		case '+':
			return Instruction.Increment;
			
		case '-':
			return Instruction.Decrement;
			
		case '.':
			return Instruction.Output;
			
		case ',':
			return Instruction.Input;
			
		case '[':
			return Instruction.TestAndJumpForward;
			
		case ']':
			return Instruction.TestAndJumpBackward;
			
		default:
			return Instruction.NOP;
		}
	}
	
}

File: src.net/BrainmessCore/Instruction.cs
using System;
using System.IO;

namespace Welch.Brainmess
{
    /// <summary>
    /// Defines all of the instructions available in Brainmess. Each type of instruction
    /// defines its own behavior given an execution context that consists of a
    /// <see cref="IProgramStream"/>, <see cref="Tape"/>, <see cref="TextReader"/> (for input),
    /// and a <see cref="TextWriter"/> ) (for output).
    /// </summary>
    public class Instruction
    {
        private readonly Action<IProgramStream, Tape, TextReader, TextWriter> _action;
        protected Instruction(Action<IProgramStream, Tape, TextReader, TextWriter> action)
        {
            _action = action;
        }

        /// <summary>
        /// Executes this instruction within the given context.
        /// </summary>
        public void Execute(IProgramStream program, Tape tape, TextReader input, TextWriter output)
        {
            _action(program, tape, input, output);
        }

        /// <summary>
        /// Returns a MoveForward instruction which when executed, moves the tape forward.
        /// </summary>
        public static readonly Instruction MoveForward =
            new Instruction((program, tape, input, output) => tape.MoveForward());

        /// <summary>
        /// Returns a MoveBackward instruction which when executed, moves the tape backward.
        /// </summary>
        public static readonly Instruction MoveBackward =
            new Instruction((program, tape, input, output) => tape.MoveBackward());

        /// <summary>
        /// Returns an Increment instruction which when executed increments the value of the current tape cell
        /// by one.
        /// </summary>
        public static readonly Instruction Increment =
            new Instruction((program, tape, input, output) => tape.Increment());

        /// <summary>
        /// Returns a Decrement instruction which when executed decrements the value of the current tape cell
        /// by one.
        /// </summary>
        public static readonly Instruction Decrement =
            new Instruction((program, tape, input, output) => tape.Decrement());

        /// <summary>
        /// Returns an Input instruction which when exectued reads one character from the input
        /// and writes its ASCII value to the tape.
        /// </summary>
        public static readonly Instruction Input =
            new Instruction((program, tape, input, output) => tape.Current = input.Read());

        /// <summary>
        /// Returns an Output instruction which when executed gets the integer value from
        /// the current tape cell, converts it to its character value and writes that to output.
        /// </summary>
        public static readonly Instruction Output =
            new Instruction((program, tape, input, output) => output.Write((char)tape.Current));

        /// <summary>
        /// Returns a TestAndJumpForward instruction. When this instruction is executed it looks at the 
        /// value of the current tape cell, and if it is non-zero it does nothing. If the value is zero,
        /// then it advances the program to the instruction following the loop.
        /// </summary>
        public static readonly Instruction TestAndJumpFoward =
            new Instruction((program, tape, input, output) => { if (tape.Current == 0) program.JumpForward(); });

        /// <summary>
        /// Returns a TestAndJumpBackward instruction. When this instruction is executed it looks at
        /// the value of the current tape cell and if it is zero it does nothing. If the value is non-zero,
        /// then it moves the tape back to the beginning of the loop.
        /// </summary>
        public static readonly Instruction TestAndJumpBackward =
            new Instruction((program, tape, input, output) => { if (tape.Current != 0) program.JumpBackward(); });

        /// <summary>
        /// Returns a NoOperation instruction which does nothing when it is executed.
        /// </summary>
        public static readonly Instruction NoOperation =
            new Instruction((program, tape, input, output) => { });

        /// <summary>
        /// Given the ascii value of a character, it returns the <see cref="Instruction"/>
        /// that matches the character.
        /// </summary>
        /// <param name="characterValue"></param>
        /// <returns></returns>
        public static Instruction FromInt(int characterValue)
        {
            switch (characterValue)
            {
                case '>':
                    return MoveForward;
                case '<':
                    return MoveBackward;
                case '+':
                    return Increment;
                case '-':
                    return Decrement;
                case '.':
                    return Output;
                case ',':
                    return Input;
                case '[':
                    return TestAndJumpFoward;
                case ']':
                    return TestAndJumpBackward;
                default:
                    return NoOperation;
            }
        }

    }


}

